### PR TITLE
[Serve] Support generators in `@serve.batch`

### LINKED ISSUE #35644: [Serve] Extend `@serve.batch` to support generator functions
### Description

The `@serve.batch` decorator can be added to a Serve deployment method to queue requests, aggregate them into microbatches, and disaggregate the responses to send them to clients. E.g. from [the docs](https://docs.ray.io/en/releases-2.4.0/serve/performance.html#enable-batching-for-your-deployment):

```python
@serve.deployment
class Model:
    @serve.batch(max_batch_size=8, batch_wait_timeout_s=0.1)
    async def __call__(self, multiple_samples: List[int]) -> List[int]:
        # Use numpy's vectorized computation to efficiently process a batch.
        return np.array(multiple_samples) * 2
```

The `@serve.batch` decorator should be extended to accept generator methods as well:

```python
@serve.deployment
class Model:
    @serve.batch(max_batch_size=8, batch_wait_timeout_s=0.1)
    async def __call__(self, multiple_samples: List[int]) -> List[int]:
        responses = [1,2,3,4]
        for response in responses:
            yield [response] * len(multiple_samples)
```

### Use case

Large language models (LLMs) stream their outputs to avoid buffering for long periods of time with no feedback. Python generators best express this type of streaming syntax, so Serve should allow batching over generators to better support LLMs.

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

When `@serve.batch` decorates a function, incoming requests are batched. The function can process the batch and return the results, and `@serve.batch` disaggregates the results and sends them back their original callers.

This change extends `@serve.batch` to support generator functions.

## Related issue number

Closes #35644

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - I added unit tests to `test_batching.py`.

### CODE REVIEW

File: python/ray/serve/batching.py [hunk 1]
```diff
@@ -141,20 +150,46 @@ async def _handle_batches(self, func):
...
-                    raise RayServeException(
-                        "Batched function doesn't preserve batch size. "
-                        f"The input list has length {len(batch)} but the "
-                        f"returned list has length {len(results)}."
-                    )
-
-                for i, result in enumerate(results):
-                    futures[i].set_result(result)
+                    func_future_or_generator = func(*args, **kwargs)
+                if isasyncgenfunction(func):
```
Comment (edoakes) [inline python/ray/serve/batching.py:157]: This function is getting really hard to follow now, can you extract the generator- and non-generator handling into separate functions?

File: python/ray/serve/batching.py [hunk 2]
```diff
@@ -141,20 +150,46 @@ async def _handle_batches(self, func):
...
+                        ]
+                        for i, result in enumerate(results):
+                            futures[i].set_result(
+                                _GeneratorIterResponse(result, new_futures[i])
+                            )
+                        futures = new_futures
+
+                    for future in futures:
+                        future.set_result(
+                            _GeneratorIterResponse(None, END_ASYNC_ITER_TOKEN)
```
Comment (edoakes) [inline python/ray/serve/batching.py:178]: Why do we need this special token? Can we have the future raise `StopIteration` instead?

File: python/ray/serve/tests/test_batching.py [hunk 3]
```diff
@@ -291,6 +292,84 @@ async def func(key1, key2):
     assert result == [("hi1", "hi2"), ("hi3", "hi4")]
 
 
+@pytest.mark.asyncio
```
Comment (edoakes) [inline python/ray/serve/tests/test_batching.py:295]: can you also add an integration test that uses this w/ `StreamingResponse`s?

It should still work without my PR merged, will just not actually stream the results back to end client iteratively.

File: python/ray/serve/batching.py [hunk 4]
```diff
@@ -13,13 +13,22 @@
...
 @dataclass
 class _SingleRequest:
     self_arg: Optional[Any]
     flattened_args: List[Any]
     future: asyncio.Future
 
 
+@dataclass
+class _GeneratorIterResponse:
+    result: Optional[Any]
```
Comment (edoakes) [inline python/ray/serve/batching.py:28]:
```suggestion
    result: Any
```
nit (these are equivalent)

**Review by edoakes — COMMENTED** [review]
Implementation looks reasonable, some stylistic comments.

File: python/ray/serve/tests/test_batching.py (hunk 3 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_batching.py:306]: I added `test_batch_generator_streaming_response_integration_test` in `test_batching.py`.

File: python/ray/serve/tests/test_batching.py [hunk 5]
```diff
@@ -291,6 +303,115 @@ async def func(key1, key2):
...
+
+    RESPONSE = "test response"
+    NUM_YIELDS = 10
+
+    @serve.deployment
+    class Textgen:
+        @serve.batch(max_batch_size=4, batch_wait_timeout_s=1000)
+        async def batch_handler(self, prompts: List[str]):
+            for _ in range(NUM_YIELDS):
+                prompt_responses = [RESPONSE] * 4
```
Comment (edoakes) [inline python/ray/serve/tests/test_batching.py:395]: add something to each yielded response to ensure they're actually returned to the correct client

maybe prepend the prompt and pass a different prompt for each call?

File: python/ray/serve/batching.py (hunk 1 above)
Comment (shrekris-anyscale) [reply python/ray/serve/batching.py]: I refactored the function and pulled some logic out to helper functions.

File: python/ray/serve/batching.py (hunk 2 above)
Comment (shrekris-anyscale) [reply python/ray/serve/batching.py]: Good idea, I changed this, so the future raises a `StopAsyncIteration` instead.

File: python/ray/serve/tests/test_batching.py (hunk 5 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_batching.py]: Good catch, I include an ID for each call now, and I check that the ID is returned to each caller.

**Comment by shrekris-anyscale** [comment]
@edoakes I've addressed all your comments. Please take a look.

File: python/ray/serve/batching.py [hunk 6]
```diff
@@ -130,34 +136,64 @@ async def wait_for_batch(self) -> List[Any]:
 
         return batch
 
-    async def _handle_batches(self, func):
+    def _validate_results(self, results, input_batch_length):
```
Comment (edoakes) [inline python/ray/serve/batching.py:139]: please add type annotations to all functions in this PR

**Review by edoakes — APPROVED** [review]
Looks great. Please add type annotations and then LGTM

File: python/ray/serve/batching.py (hunk 6 above)
Comment (shrekris-anyscale) [reply python/ray/serve/batching.py]: I added type annotations.

**Comment by shrekris-anyscale** [comment]
The test failures are unrelated:

1. Windows Build & Test: [`test_task_events`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/23510#01885517-2028-421c-a314-e95618640b63/2686-3239) is broken on `master`:

Screen Shot 2023-05-25 at 9 36 46 PM

2. Windows Wheels: [error](https://buildkite.com/ray-project/oss-ci-build-pr/builds/23510#01885517-202b-4393-bfbc-be5de3bba0db/4161-4180) seems to be an unrelated issue with `npm`:

```console
npm ERR! code EEXIST
--
  | npm ERR! syscall copyfile
  | npm ERR! path C:\Users\ContainerAdministrator\AppData\Local\npm-cache\_cacache\tmp\a07abb13
  | npm ERR! dest C:\Users\ContainerAdministrator\AppData\Local\npm-cache\_cacache\content-v2\sha512\6f\7f\5305a4d27d5eb206b6a953cf69e5f29e904da6fcdc270e870e56bb90152d7fbde320773b8f72738cdf833a0b0c56f231ff97111ae6b0680de530bb91c74f
  | npm ERR! errno -4082
  | npm ERR! EBUSY: resource busy or locked, copyfile 'C:\Users\ContainerAdministrator\AppData\Local\npm-cache\_cacache\tmp\a07abb13' -> 'C:\Users\ContainerAdministrator\AppData\Local\npm-cache\_cacache\content-v2\sha512\6f\7f\5305a4d27d5eb206b6a953cf69e5f29e904da6fcdc270e870e56bb90152d7fbde320773b8f72738cdf833a0b0c56f231ff97111ae6b0680de530bb91c74f'
  | npm ERR! File exists: C:\Users\ContainerAdministrator\AppData\Local\npm-cache\_cacache\content-v2\sha512\6f\7f\5305a4d27d5eb206b6a953cf69e5f29e904da6fcdc270e870e56bb90152d7fbde320773b8f72738cdf833a0b0c56f231ff97111ae6b0680de530bb91c74f
  | npm ERR! Remove the existing file and try again, or run npm
  | npm ERR! with --force to overwrite files recklessly.
```

3. Operator: [error](https://buildkite.com/ray-project/oss-ci-build-pr/builds/23510#0188551f-dddc-4fe4-bd87-391a89cf12de/1077-1150) seems to be unrelated issue with KubeRay:

```console
subprocess.CalledProcessError: Command '['kubectl', 'delete', '--ignore-not-found', '-k', 'ray/python/ray/autoscaler/kuberay/config/default']' returned non-zero exit status 1.
```

**Comment by shrekris-anyscale** [comment]
@edoakes @architkulkarni This change is ready to merge.

File: python/ray/serve/tests/test_batching.py [hunk 7]
```diff
@@ -291,6 +296,113 @@ async def func(key1, key2):
...
+        elif error_type == "mismatched_lengths":
+            with pytest.raises(RayServeException):
+                await generator.__anext__()
+
+        with pytest.raises(StopAsyncIteration):
+            await generator.__anext__()
+
+
+@pytest.mark.asyncio
+async def test_batch_generator_streaming_response_integration_test(serve_instance):
```
Comment (akshay-anyscale) [inline python/ray/serve/tests/test_batching.py:378]: Should we have a test where the two outputs are of different lengths?

Comment (akshay-anyscale) [reply python/ray/serve/tests/test_batching.py:378]: Where different outputs in the batch are of different lengths (as we discussed offline).
